### ISSUE TITLE
building a code that utilizes Numba and mpi4py 

### ISSUE THREAD

**Issue by ahmad681** [numba/numba#8692]
Hi, I don't think this has been addressed before. I am trying to simulate atomic diffusion using Monte Carlo method. First I am decomposing the domain size into subdomains and operate on each subdomain in different processors. Of course, a lot of of MPI.Send() and MPI.Recv() communications are required for my problem. I don't use numba in functions that call send and recv because I don't think Numba supports that and there is no reason to do so i believe. 
Here is my problem. So I make buffers per every subdomain with unique sizes, and Send() them to nearby processors so that they can Recv() it. Then I use Numba to act on specific region of the subdomain in addition to the buffers that a processor has just received. My code looks as following:

#quad 0
#@njit
def get_state_quad_a(ind_1st_nei, z_current):
    output = np.zeros((ind_1st_nei.shape[0],ind_1st_nei.shape[1]), dtype=np.ubyte)
    
    for atom in range(len(ind_1st_nei)): 
        for nei in range(24): 
            if ((ind_1st_nei[atom][nei][0] < 0) and (ind_1st_nei[atom][nei][1] < 0)):
                output[atom][nei] = (buff_corn[ind_1st_nei[atom][nei][0] + NUMBER_NEIGHBOR, ind_1st_nei[atom][nei][1] + NUMBER_NEIGHBOR, (ind_1st_nei[atom][nei][2] - z_current + 1)]) 
            elif((ind_1st_nei[atom][nei][0] < 0) and (ind_1st_nei[atom][nei][1] > 0)): 
                output[atom][nei] = (buff_vert[ind_1st_nei[atom][nei][0] + NUMBER_NEIGHBOR, ind_1st_nei[atom][nei][1], (ind_1st_nei[atom][nei][2] - z_current + 1)]) 
            elif((ind_1st_nei[atom][nei][0] > 0) and (ind_1st_nei[atom][nei][1] < 0)): 
                output[atom][nei] = (buff_horz[ind_1st_nei[atom][nei][0], ind_1st_nei[atom][nei][1] + NUMBER_NEIGHBOR, (ind_1st_nei[atom][nei][2] - z_current + 1)]) 
            else: 
                output[atom][nei] = State[ind_1st_nei[atom][nei][0],ind_1st_nei[atom][nei][1],ind_1st_nei[atom][nei][2]]
    return output

So basically this function reads the information of ind_1st_nei that is a 3D array (physically this is the location of sites in the array to determine location of atoms). And I use that array to read the State (also 3D array) in the current step to tell me what type of atom there is there (0,1,2,3...etc). Passing some conditions (near subdomain boundaries), I need to read the buffers (buff_vert, buff_corn and buff_horz). 
When I don't njit, the results come out fine, I read correct neighbors states, however, when I use numba, at some point, the simulations messes up, the output array comes out in correct, it is hard to predict where the error is. 
I am afraid there's a race condition happening, but i believe this should not be the case, because every processor should technically have its own memory. Any suggestions? thanks a lot.  

## Reporting a bug

- [ ] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [ ] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

**Comment by apmasell** [comment]
Numba doesn't directly support MPI communication, but there is a package that allows MPI functions to be used in JITed code: https://pypi.org/project/numba-mpi/

Have you looked at a single step of the simulation? That might give more insight into where the problem is.

**Comment by ahmad681** [comment]
I am not using numba on any function that calls Send or Recv. 
I am using numba on arrays including buffer array that I should in theory work on every processor. 
I did look at a single step here what I can tell: 
so the output array with using numba njit looks like the following: 
array([[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
        3, 3]], dtype=uint8)

without using numba njit, I see as expected the following array: 
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
        3, 3]], dtype=uint8)

So not sure where it is writing 1's in the above array when using numba. I feel like it is overwriting things by accessing incorrect State or buff_   ndarray! Completely lost at this point. thanks a lot for your reply.

**Comment by ahmad681** [comment]
Okay here's a small update if that can help. So I am running things in debug mode, 
before setting up a breaking point within get_state_quad_a(ind_1st_nei, z_current) function, I set up a breaking point on the statement that calls this function. When I step in, it takes me to this: 
/.conda/envs/kmc/lib/python3.9/site-packages/numba/core/serialize.py(29)_numba_unpickle()
-> def _numba_unpickle(address, bytedata, hashed):
> /home/ahmad68/.conda/envs/kmc/lib/python3.9/site-packages/numba/core/serialize.py(43)_numba_unpickle()
-> key = (address, hashed)
(Pdb) n
> /home/ahmad68/.conda/envs/kmc/lib/python3.9/site-packages/numba/core/serialize.py(44)_numba_unpickle()
-> try:
(Pdb) n
> /home/ahmad68/.conda/envs/kmc/lib/python3.9/site-packages/numba/core/serialize.py(45)_numba_unpickle()
-> obj = _unpickled_memo[key]
(Pdb) n
> /home/ahmad68/.conda/envs/kmc/lib/python3.9/site-packages/numba/core/serialize.py(48)_numba_unpickle()
-> return obj
(Pdb) n
--Return--
> /home/ahmad68/.conda/envs/kmc/lib/python3.9/site-packages/numba/core/serialize.py(48)_numba_unpickle()-><class 'numpy.ndarray'>
-> return obj

I don't know what this means quite frankly.

**Comment by ahmad681** [comment]
quick update: I tried to debug the decorated function with njit but I don't think this is allowed with mpi environment. 
I thought perhaps numba doesn't work with "and" statements, so I tried using np.logical_and, and I still get same numerical issue. 
I believe it is either racing condition which it shouldn't happen, or that the njit is not working with its own Cache! 
I would really appreciate the help in this situation, becuase numba would make my simulation at least 6 times faster based on initial code testing. many thanks in advance.

**Comment by apmasell** [comment]
Okay, just to be clear, you can reproduce the issue without MPI? If so, can you produce a complete script that will show the error?

`_numba_unpickle` is used where Numba will convert certain bits of data into the machine code it generates and has to convert it back. You're mostly seeing Numba internals there.

**Comment by ahmad681** [comment]
actually, without MPI (no domain decomposition) and running serial code with few decorated function in numba like the following, the code runs just fine. 
@njit
def getFNThreaded(occupied): 
    first_nei_index = [[ind_conn(atom_occupied, i, n) for i in range(24)] for atom_occupied in occupied]
    return first_nei_index

this function technically takes in an 3D array where occupied sites are located, and I loop through them to deterime the nearby 24 nearest sites. 
This function works great with numba. 
I have implemented the same version of it in the paralleize MPI code, and this function particularly works fine as well. It is just the function above that is spitting out good results with numba.

**Comment by ahmad681** [comment]
@apmasell, let me ask you this: 
the buffers being sent and received between processors "buff_vert, buff_horz and buff_corn" are actually in theory contiguous arrays. 
Technically, these arrays are copies from an array from nearby processor and being sent in contiguous fation using np.ascontiguousarray() 
So most likely the indexing will be different, do you think that is a problem when using numba? clearly i am facing an indexing issue I would believe so. 
Thanks a lot

**Comment by stuartarchibald** [comment]
@ahmad681 does the _exact_ same "numba" code work when all the `@njit` decorators are removed (you can also just set the env var `NUMBA_DISABLE_JIT=1` to achieve the same). Am asking just to see if the code is correct when used in the interpreter with pure NumPy.

**Comment by ahmad681** [comment]
yes, the code works just fine without @njit, and the output is correct. 
With @njit, no errors, code still works, but the output is incorrect

**Comment by stuartarchibald** [comment]
@ahmad681 thanks for confirming. It's potentially going to be hard to differentiate between "Issue in the MPI implementation" and "bug in Numba" as the only reported observable side effect so far is that the answer is sometimes "wrong".

Some suggestions:
* Perhaps use the no-numba version and try using one or two MPI processes and see what effect that has, then do the same with a large number of MPI processes. Maybe if there's a bug in the message passing logic or if there's a race condition, adjusting these things might make it show up.
* Do as above but with Numba present, what happens?
* Try shrinking the code down by removing logic until the no-numba and numba versions produce the same output consistently, that might show where the problem is.

Hope this helps? I should also add that, without a minimal working reproducer, debugging this is unfortunately going to be quite hard.

**Comment by ahmad681** [comment]
@stuartarchibald  thanks a lot for the suggestions, much appreciated. I spent my entire day debugging the output. Here's an interesting result, numba apperantly is messing things up with the if( (condition and condition ) statements, there has to be a way to go around this, but I am not sure how. Check out the following output.  so I added the following print statements: 

@njit
def get_state_quad_a(ind_1st_nei, z_current):
    output = np.empty((ind_1st_nei.shape[0],ind_1st_nei.shape[1]), dtype=np.ubyte)
    print("this is inside numba \n",ind_1st_nei[0])
    for atom in range(len(ind_1st_nei)): 
        if (atom == 0):
                print("indexing atom =0 ", "becomes",ind_1st_nei[atom])
        for nei in range(24): 
            if (ind_1st_nei[atom,nei,0] < 0 and ind_1st_nei[atom,nei,1] < 0):
                output[atom,nei] = buff_corn[ind_1st_nei[atom,nei,0] + NUMBER_NEIGHBOR, ind_1st_nei[atom,nei,1] + NUMBER_NEIGHBOR, (ind_1st_nei[atom,nei,2] - z_current + 1)] 
            elif(ind_1st_nei[atom,nei,0] < 0 and ind_1st_nei[atom,nei,1] > 0): 
                output[atom,nei] = buff_vert[ind_1st_nei[atom,nei,0] + NUMBER_NEIGHBOR, ind_1st_nei[atom,nei,1], (ind_1st_nei[atom,nei,2] - z_current + 1)] 
            elif(ind_1st_nei[atom,nei,0] > 0 and ind_1st_nei[atom,nei,1] < 0): 
                output[atom,nei] = buff_horz[ind_1st_nei[atom,nei,0], ind_1st_nei[atom,nei,1] + NUMBER_NEIGHBOR, (ind_1st_nei[atom,nei,2] - z_current + 1)] 
            else: 
                output[atom,nei] = State[ind_1st_nei[atom,nei,0],ind_1st_nei[atom,nei,1],ind_1st_nei[atom,nei,2]]     
                if (atom == 0):
                    print("at atom =", atom,"nei is", nei, "wher first_nei_index is", ind_1st_nei[atom,nei,:], "\n", "and state is",State[ind_1st_nei[atom,nei,0],ind_1st_nei[atom,nei,1],ind_1st_nei[atom,nei,2]])       
  #  gdb_breakpoint()
    return output

and the output comes out from the print statement
this is inside numba [[ 1  7  5]
 [ 3  5  5]
 [-1  5  5]
 [ 1  3  5]
 [ 3  7  5]
 [-1  7  5]
 [ 3  3  5]
 [-1  3  5]
 [ 1  9  5]
 [ 5  5  5]
 [ 1  1  5]
 [-3  5  5]
 [ 2  6  4]
 [ 2  4  4]
 [ 0  6  4]
 [ 0  4  4]
 [ 2  8  4]
 [ 2  2  4]
 [ 0  8  4]
 [ 0  2  4]
 [ 4  6  4]
 [ 4  4  4]
 [-2  6  4]
 [-2  4  4]]
indexing atom =0  becomes [[ 1  7  5]
 [ 3  5  5]
 [-1  5  5]
 [ 1  3  5]
 [ 3  7  5]
 [-1  7  5]
 [ 3  3  5]
 [-1  3  5]
 [ 1  9  5]
 [ 5  5  5]
 [ 1  1  5]
 [-3  5  5]
 [ 2  6  4]
 [ 2  4  4]
 [ 0  6  4]
 [ 0  4  4]
 [ 2  8  4]
 [ 2  2  4]
 [ 0  8  4]
 [ 0  2  4]
 [ 4  6  4]
 [ 4  4  4]
 [-2  6  4]
 [-2  4  4]]
at atom = 0 nei is 0 wher first_nei_index is [1 7 5] 
 and state is 0
at atom = 0 nei is 1 wher first_nei_index is [3 5 5] 
 and state is 0
at atom = 0 nei is 3 wher first_nei_index is [1 3 5] 
 and state is 0
at atom = 0 nei is 4 wher first_nei_index is [3 7 5] 
 and state is 0
at atom = 0 nei is 6 wher first_nei_index is [3 3 5] 
 and state is 0
at atom = 0 nei is 8 wher first_nei_index is [1 9 5] 
 and state is 0
at atom = 0 nei is 9 wher first_nei_index is [5 5 5] 
 and state is 0
at atom = 0 nei is 10 wher first_nei_index is [1 1 5] 
 and state is 0
at atom = 0 nei is 12 wher first_nei_index is [2 6 4] 
 and state is 3
at atom = 0 nei is 13 wher first_nei_index is [2 4 4] 
 and state is 3
at atom = 0 nei is 14 wher first_nei_index is [0 6 4] 
 and state is 3
at atom = 0 nei is 15 wher first_nei_index is [0 4 4] 
 and state is 3
at atom = 0 nei is 16 wher first_nei_index is [2 8 4] 
 and state is 3
at atom = 0 nei is 17 wher first_nei_index is [2 2 4] 

if you notice, in ind_1st_nei[0,2,:] should read like  [-1  5  5], but when it hits the if statement, it should still read [-1,5,5] but I access data from buff_vert[-1+6,5,5] which is data point. but once the for loop has been executed, the output as you can see is [1 3 5], and so I am not sure where these numbers are coming from quite frankly. I thought about sharing this with the community maybe you might a have an explanation of this. Note: without numba, it doesn't do that weird output. 

I am still continuing working around this and I'll keep updating. thanks a lot numba team. :)

**Comment by ahmad681** [comment]
Thanks a lot for the suggestion @stuartarchibald  
I fixed the problem by doing the following, and it worked. Apperantely there was a problem in storing the State[:,:,:] ndarray. when i was printing this array with and without njit, the arrays were completely different. for some reason when using njit, this State[:,:,:] was probably over written from somewhere that I couldn't tell, or the inside numba function the nature of State[:,:,:] was not being read correctly. Long story short, all i did is that I passed State as a variable, and allocated new memory for it using .copy(), and it worked just fine.
